File: leetcode1717/solution.cpp
class Solution {
public:
    int maximumGain(string s, int x, int y) {
        /*
        Time: O(n) | Space: O(n)

        where 'n' is the length of 's'
        */
        vector<char> chars;

        int res = 0;

        for (int i = 0; i < s.size(); ++i) {
            chars.push_back(s[i]);
        }

        if (x > y) {
            chars = helper(chars, true, x, &res);
            chars = helper(chars, false, y, &res);
        } else {
            chars = helper(chars, false, y, &res);
            chars = helper(chars, true, x, &res);
        }

        return res;


    }

    vector<char> helper(vector<char> chars, bool x, int step, int* sum) {
        vector<char> ret;

        if (x) {
            for (int i = 0; i < chars.size(); ++i) {
                if (!ret.empty() && ret.back() == 'a' && chars[i] == 'b') {
                    ret.pop_back();
                    *sum += step;
                } else {
                    ret.push_back(chars[i]);
                }
            }
        } else {
            for (int i = 0; i < chars.size(); ++i) {
                if (!ret.empty() && ret.back() == 'b' && chars[i] == 'a') {
                    ret.pop_back();
                    *sum += step;
                } else {
                    ret.push_back(chars[i]);
                }
            }
        }

        return ret;
    }
};

File: leetcode1717/solution.rs
impl Solution {
    pub fn maximum_gain(s: String, x: i32, y: i32) -> i32 {
        /// Time: O(n) | Space: O(n)
        /// 
        /// where 'n' is the length of 's'
        let mut chars: Vec<char> = s.chars().collect::<Vec<char>>();
        let mut res: i32 = 0;

        if x > y {
            chars = Self::helper(&chars, true, x, &mut res);
            chars = Self::helper(&chars, false, y, &mut res);
        } else {
            chars = Self::helper(&chars, false, y, &mut res);
            chars = Self::helper(&chars, true, x, &mut res);
        }

        res
    }

    fn helper(chars: &Vec<char>, x: bool, step: i32, sum: &mut i32) -> Vec<char> {
        let mut ret: Vec<char> = Vec::new();

        if x {
            for &c in chars.iter() {
                if !ret.is_empty() && *ret.last().unwrap() == 'a' && c == 'b' {
                    ret.pop();
                    *sum += step;
                } else {
                    ret.push(c);
                }
            }
        } else {
            for &c in chars.iter() {
                if !ret.is_empty() && *ret.last().unwrap() == 'b' && c == 'a' {
                    ret.pop();
                    *sum += step;
                } else {
                    ret.push(c);
                }
            }
        }

        ret
    }
}